codec/src/main/java/io/netty/handler/codec/base64/Base64.java

```diff
@@ -222,93 +301,109 @@ public static ByteBuf decode(
+            int len34 = (len * 3) >>> 2;
+            dest = allocator.buffer(len34).order(src.order()); // Upper limit on size of output
+
+            decodabet = decodabet(dialect);
+            try {
+                src.forEachByte(off, len, this);
+                ByteBuf slice = dest.slice(0, outBuffPosn);
+                return slice;
+            } catch (Throwable cause) {
+                dest.release();
+                PlatformDependent.throwException(cause);
+                return null;
+            }
+        }
+
+        @Override
+        public boolean process(byte value) throws Exception {
+            sbiCrop = (byte) (value & 0x7f); // Only the low seven bits
+            sbiDecode = decodabet[sbiCrop];
 
             if (sbiDecode >= WHITE_SPACE_ENC) { // White space, Equals sign or better
                 if (sbiDecode >= EQUALS_SIGN_ENC) { // Equals sign or better
                     b4[b4Posn ++] = sbiCrop;
                     if (b4Posn > 3) { // Quartet built
-                        outBuffPosn += decode4to3(
-                                b4, 0, dest, outBuffPosn, dialect);
+                        outBuffPosn += decode4to3(b4, dest, outBuffPosn, decodabet);
                         b4Posn = 0;
 
                         // If that was the equals sign, break out of 'for' loop
                         if (sbiCrop == EQUALS_SIGN) {
-                            break;
+                            return false;
                         }
                     }
                 }
-            } else {
-                throw new IllegalArgumentException(
-                        "bad Base64 input character at " + i + ": " +
-                        src.getUnsignedByte(i) + " (decimal)");
+                return true;
             }
+            throw new IllegalArgumentException(
+                    "invalid bad Base64 input character: " + (short) (value & 0xFF) + " (decimal)");
         }
 
-        return dest.slice(0, outBuffPosn);
-    }
+        private static int decode4to3(byte[] src, ByteBuf dest, int destOffset, byte[] decodabet) {
+            if (src[2] == EQUALS_SIGN) {
+                // Example: Dk==
+                int outBuff = (decodabet[src[0]] & 0xFF) << 18 |
+                              (decodabet[src[1]] & 0xFF) << 12;
+
+                dest.setByte(destOffset, (byte) (outBuff >>> 16));
+                return 1;
+            }
+
+            if (src[3] == EQUALS_SIGN) {
+                // Example: DkL=
+                int outBuff = (decodabet[src[0]] & 0xFF) << 18 |
+                              (decodabet[src[1]] & 0xFF) << 12 |
+                              (decodabet[src[2]] & 0xFF) << 6;
+
+                // Packing bytes into a short to reduce bound and reference count checking.
+                if (dest.order() == ByteOrder.BIG_ENDIAN) {
+                    dest.setShort(destOffset,
```

Comment:
you could also get rid of the `outBuff` and just directly put the bytes where they need to be